### PR TITLE
[Refactoring] Add first multi-participant on-device tests, further refactoring of the CallingSDK interface

### PR DESCRIPTION
**Pull request by grigoryk**
## Purpose
This PR adds a first set of multi-participant UI tests that are written against an in-memory implementation of the Calling SDK. The tests are able to programmatically invoke remote events, such as adding or removing remote participants from the call, and then verify that proper UI state is rendered.

To achieve this, several steps were taken:
- We further refactored the codebase (continuing from https://github.com/Azure/communication-ui-library-android/pull/367) to hide all the rest of relevant native library SDK bits behind an interface
- Added a mechanism for injecting custom implementations of the CallingSDK into the app (via shared global state)
- Updated `AppStore` to use a single-threaded coroutine dispatcher, in order to allow its injection from the test environment (TODO: verify that this is actually necessary to get tests sequenced correctly; I think so?)
- Finally, added several UI tests that instantiate an in-memory CallingSDK, inject it into the library before launching the composite, and then simulate and verify various remote events

Best reviewed by commits, as they form a nice logical sequence of changes.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Pull Request Type
What kind of change does this Pull Request introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes
- [x] Other... Please describe:
First set of on-device UI tests against a headless in-memory SDK.
## Pull Request Checklist

- [ ] Public API changes
- [ ] Verified for cross-platform
- [ ] Synced with iOS, Web team
- [ ] Internal review completed
- [ ] UI Changes
  - [ ] Screen captures included
  - [ ] Tested for Light/Dark mode
  - [ ] Tested for screen rotation
  - [ ] Tested on Tablet and small screen device (5")
  - [ ] Include localization changes
- [ ] Tests
  - [ ] Memory leak analysis performed
  - [ ] Tested on API 21, 26 and latest 
  - [ ] Tested for background mode
  - [ ] Unit Tests Included
  - [ ] UI Automated Tests Included

## How to Test

* Open...
* Click on...

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

**Comment by grigoryk** [comment]
Now, this is probably where something like Rust's traits would have really helped - being able to implement interfaces for classes we don't own! But working with what we have...
 
I'm not really a fan of the use of generics within CallingSDK interface, but not sure yet if they can be avoided in this approach (although definitely cleaned up from the current state).

For example: for both `RemoteVideoStream` and `LocalVideoStream`, we need to be able to "convert" back into their native counter-parts when passing them into native libs - for rendering and starting/stopping local video. So, in this current pass of this refactor, I box these native objects using wrapper classes that implement corresponding local interfaces. Use of generics lets us actually define the specific underlying types we're really working with. Could have used `Any` instead to simplify things, but we'd loose the type safety then.

For testing purposes, this approach lets us fully avoid having to touch native libs entirely.

**Comment by grigoryk** [comment]
The next step here is to write a simple multi-participant test case against an in-memory implementation of the `CallingSDK` interface. We'll use that to inform this final shape of this refactor and then work toward landing the two alongside each other. @iaulakh will work on the test.

**Comment by grigoryk** [comment]
All right, this is in a decent shape now. Two TODOs remaining:
- update AppStore unit tests after the conversion away from raw Handler
- verify that indeed, having control over AppStore dispatcher is necessary for these tests. It isn't necessary for unit tests, so what's different about the instrumented tests?

**Comment by grigoryk** [comment]
CI is green other than an unrelated failure on `SCC-SPOOL-communication-ui-library-android-Azure-Monitor` pipeline; that's failing since `CallingTelemetryDebug` build variant isn't there. Seems like the pipeline got ahead of `develop` a little bit - that work is present on the [feature/event_log](https://github.com/Azure/communication-ui-library-android/tree/feature/event_log) branch.

**Comment by iaulakh** [comment]
seeing crash: will investigate soon
Process: com.azure.android.communication.ui.callingcompositedemoapp.debug, PID: 3605
    com.azure.android.communication.calling.CallingCommunicationException
        at com.azure.android.communication.calling.VideoStreamRenderer.<init>(VideoStreamRenderer.java:77)
        at com.azure.android.communication.calling.VideoStreamRenderer.<init>(VideoStreamRenderer.java:60)
        at com.azure.android.communication.ui.calling.service.sdk.VideoStreamRendererLocalWrapper.createView(VideoStreamRendererLocalWrapper.kt:21)
        at com.azure.android.communication.ui.calling.presentation.VideoViewManager.updateLocalVideoRenderer(VideoViewManager.kt:70)
        at com.azure.android.communication.ui.calling.presentation.fragment.calling.localuser.LocalParticipantView.setLocalParticipantVideo(LocalParticipantView.kt:203)
        at com.azure.android.communication.ui.calling.presentation.fragment.calling.localuser.LocalParticipantView.access$setLocalParticipantVideo(LocalParticipantView.kt:25)
        at com.azure.android.communication.ui.calling.presentation.fragment.calling.localuser.LocalParticipantView$start$1$invokeSuspend$$inlined$collect$1.emit(Collect.kt:135)
        at kotlinx.coroutines.flow.StateFlowImpl.collect(StateFlow.kt:398)
        at kotlinx.coroutines.flow.StateFlowImpl$collect$1.invokeSuspend(Unknown Source:15)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTaskKt.resume(DispatchedTask.kt:234)
        at kotlinx.coroutines.DispatchedTaskKt.resumeUnconfined(DispatchedTask.kt:190)
        at kotlinx.coroutines.DispatchedTaskKt.dispatch(DispatchedTask.kt:161)
        at kotlinx.coroutines.CancellableContinuationImpl.dispatchResume(CancellableContinuationImpl.kt:397)
        at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl(CancellableContinuationImpl.kt:431)
        at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl$default(CancellableContinuationImpl.kt:420)
        at kotlinx.coroutines.CancellableContinuationImpl.resumeWith(CancellableContinuationImpl.kt:328)
        at kotlinx.coroutines.flow.StateFlowSlot.makePending(StateFlow.kt:286)
        at kotlinx.coroutines.flow.StateFlowImpl.updateState(StateFlow.kt:351)
        at kotlinx.coroutines.flow.StateFlowImpl.setValue(StateFlow.kt:318)
        at com.azure.android.communication.ui.calling.presentation.fragment.calling.localuser.LocalParticipantViewModel.update(LocalParticipantViewModel.kt:57)
        at com.azure.android.communication.ui.calling.presentation.fragment.calling.CallingViewModel.onStateChange(CallingViewModel.kt:151)
        at com.azure.android.communication.ui.calling.presentation.fragment.BaseViewModel$init$1$invokeSuspend$$inlined$collect$1.emit(Collect.kt:135)

**Comment by grigoryk** [comment]
Latest push fixes issues around camera switching.

**Comment by iaulakh** [comment]
@pavelprystinka @Mohtasim can you lease review?

**Review by iaulakh — DISMISSED** [review]

**Review by Mohtasim — DISMISSED** [review]

File: azure-communication-ui/azure-communication-ui/build.gradle [hunk 1]
```diff
@@ -142,4 +142,8 @@ dependencies {
 
     androidTestImplementation "androidx.test.ext:junit:$androidx_junit_version"
     androidTestImplementation "androidx.test.espresso:espresso-core:$androidx_espresso_core_version"
+    androidTestImplementation 'androidx.test:rules:1.4.1-alpha07'
+    androidTestImplementation "androidx.test.espresso:espresso-core:$androidx_espresso_core_version"
```
[the remark this answers is not in the document]
Comment (grigoryk) [reply azure-communication-ui/azure-communication-ui/build.gradle:146]: Thanks for catching!

**Review by ahammer — APPROVED** [review]

**Review by Mohtasim — APPROVED** [review]

**Pull request merged**